src/validation.cpp

```diff
@@ -2219,28 +2232,40 @@ class ConnectTrace {
     }
 
     void BlockConnected(CBlockIndex* pindex, const std::shared_ptr<const CBlock>& pblock) {
-        blocksConnected.emplace_back(pindex, pblock);
-        conflictedTxs.emplace_back();
+        assert(!awaitingClear);
+        assert(pindex);
+        assert(pblock);
+        blocksConnected.back().pindex = pindex;
+        blocksConnected.back().pblock = pblock;
+        blocksConnected.emplace_back();
+    }
+
+    std::vector<PerBlockConnectTrace>& GetBlocksConnected() {
+        // We always keep one extra block at the end of our list because
+        // blocks are added after all the conflicted transactions have
+        // been filled in. Thus, the last entry should always be an empty
+        // one waiting for the transactions from the next block. We use
+        // awaitingClear to keep track of this state, and pop the last
+        // entry here to make sure the list we return is sane.
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
```

Comment:
In commit "Handle SyncTransaction in ActivateBestChain instead of ConnectTrace"

Could / should you `assert(blocksConnected.back().conflictedTxs.empty())` here?